### PR TITLE
tests/unittests/tests-core-clist: improve unit test

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- Iterate of different lengths of unsorted data to also cover corner case one-item list
- Actually check that the list is sorted afterwards (and not just that the maximum got sorted in last)
- Actually check that the list was stable-sorted (as the API doc promises a stable sort)

### Testing procedure

- The unit tests should pass
- Mess with the sorting algorithm, check that the test stops passing

### Issues/PRs references

None

### CODE REVIEW

File: tests/unittests/tests-core/tests-core-clist.c [hunk 1]
```diff
@@ -266,19 +279,80 @@ static void test_clist_foreach(void)
...
 }
 
-static int _cmp(clist_node_t *a, clist_node_t *b)
+static int _cmp(clist_node_t *_a, clist_node_t *_b)
 {
-    /* this comparison function will sort by the actual memory address of the
-     * list node (descending) */
-    return (uintptr_t)a - (uintptr_t) b;
+    struct test_list_node *a = _get_test_list_node(_a), *b = _get_test_list_node(_b);
+    return a->value - b->value;
```
Comment (crasbe) [inline tests/unittests/tests-core/tests-core-clist.c:285]: Isn't this an implicit type conversion? The `value` member of the list is `unsigned` and the return value is `int`.

File: tests/unittests/tests-core/tests-core-clist.c [hunk 2]
```diff
@@ -266,19 +279,80 @@ static void test_clist_foreach(void)
...
+        0x2a,   0xc6,   0xd3,   0x55,   0x62,   0xa2,   0xca,   0x5c,
+        0xeb,   0xfe,   0x4e,   0x64,   0xfe,   0xb2,   0x34,   0xc2,
+        0xa8,   0xdd,   0xe9,   0x5d,   0x1b,   0x6c,   0xd2,   0xa2,
+        0x66,   0xe6,   0x10,   0x81,   0xb2,   0xab,   0x59,   0xe3,
+        0x67,   0x67,   0xcd,   0xbc,   0xcc,   0x4f,   0xa9,   0x04,
+        0xd5,   0x5a,   0x98,   0x1e,   0x75,   0x3f,   0xf5,   0x2b,
+        0xcb,   0xda,   0x50,   0xad,   0xb0,   0xeb,   0xea,   0x9a,
+    };
+    static_assert(ARRAY_SIZE(tests_clist_buf) <= ARRAY_SIZE(values),
+                  "Not enough test date for sort test");
```
Comment (crasbe) [inline tests/unittests/tests-core/tests-core-clist.c:348]:
```suggestion
                  "Not enough test data for sort test");
```

Typo?

File: tests/unittests/tests-core/tests-core-clist.c [hunk 3]
```diff
@@ -295,22 +369,11 @@ static void test_clist_sort(void)
...
-
-    /* sort list */
-    clist_sort(list, _cmp);
-
-    uintptr_t last = (uintptr_t) list->next;
-
-    for (int i = 0; i < TEST_CLIST_LEN; i++) {
-        TEST_ASSERT((uintptr_t) clist_rpop(list) <= last);
+    for (size_t cur_len = 0; cur_len < TEST_CLIST_LEN; cur_len++) {
+        _prepare_unsorted_clist(cur_len);
```
Comment (crasbe) [inline tests/unittests/tests-core/tests-core-clist.c:373]: Github doesn't let me select the according lines, but the documentation has to be adapted as well. It currently says `(using clist_poprpush())` still.

**Review by crasbe — COMMENTED** [review]
Disclaimer: I still don't know anything about `clist` :D

There are also two static errors about line length.

File: tests/unittests/tests-core/tests-core-clist.c [hunk 4]
```diff
@@ -266,51 +283,106 @@ static void test_clist_foreach(void)
...
+
+    clist_node_t *end_of_clist = list->next;
+    clist_node_t *prev = list->next->next;
+    clist_node_t *cur = prev->next;
+
+    while (prev != end_of_clist) {
+        struct test_list_node *a = _get_test_list_node(prev);
+        struct test_list_node *b = _get_test_list_node(cur);
+
+        if (a->value > b->value) {
```
[the remark this answers is not in the document]
Comment (maribu) [reply tests/unittests/tests-core/tests-core-clist.c]: Done

**Review by benpicco — APPROVED** [review]
Please squash

**Comment by maribu** [comment]
Thx!

File: tests/unittests/tests-core/tests-core-clist.c (hunk 1 above)
Comment (maribu) [reply tests/unittests/tests-core/tests-core-clist.c:290]: I changed the value of the `struct test_list_node::value` to `int` to avoid the implicit conversion.

Comment (crasbe) [reply tests/unittests/tests-core/tests-core-clist.c:290]: Very good :)
